### PR TITLE
Improve documentation

### PR DESCRIPTION
**Pull request by OCHyams**
Update README.md - it was fairly out of date.

Add Command.md - document which covers Dexter commands for users (test writers)

Improve docstrings - Use "Google style" ish docstrings to better document as
much of Dexter as possible.

There is some very minor code clean up in places which could help self
documentation, eg function names.

Note:
I've written this as if PR #33 and #37 have landed.

### CODE REVIEW

File: dex/command/commands/DexExpectProgramState.py [hunk 1]
```diff
@@ -40,6 +40,13 @@ def state_from_dict(source: dict) -> ProgramState:
     return ProgramState(**source)
 
 class DexExpectProgramState(CommandBase):
+    """Expect to see a given program `state` a certain numer of `times`.
+
+    DexExpectProgramState(state [,**times])
+
+    See Commands.md for more info.
+    """
```
Comment (OCHyams) [inline dex/command/commands/DexExpectProgramState.py]: My thought processes here is that the commands are not used in Dexter itself, so better to have most of the documentation in Commands.md and reduce the changes of mismatching info.

File: Commands.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                            'column': int,
+                          },
+                          'watches': {
+                            expr (str): value (str),
+                          },
+                        }
+                      ]}
+
+    Keyword args:
+        times (int): Minumum number of ime this state pattern is expected to be
```
Comment (Melamoto) [inline Commands.md]: typo

File: Commands.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            seen. Defaults to 1. Can be 0.
+
+### Description
+Expect to see a given program `state` a certain numer of `times`.
```
Comment (Melamoto) [inline Commands.md]: typo

File: Commands.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+When matching expected against reported states:
+* Omitted fields in `StackFrame` and `SourceLocation` dictionaries are ignored.
+* The state match fails if the results of each `expr` in `watches`
```
Comment (Melamoto) [inline Commands.md]: Location is matched as well (also the function name and inline status should probably be matched, but that's a change for the future).

File: Commands.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+## DexLabel
+    DexLabel(name)
+
+    Args:
+        name (str): A unique name for this line.
+
+### Description
+Name the line this command is found on. Line names can be referenced by other
+commands expecting line number arguments.
+For example, `DexExpectWatchValues(..., on_line='my_line_name')`.
```
[the remark this answers is not in the document]
Comment (OCHyams) [reply Commands.md]: From a user standpoint I think we should fix this. You'd expect labels to work in both DexExpectProgramState and DexExpectWatchValues, no?

File: Commands.md (hunk 4 above)
Comment (OCHyams) [reply Commands.md]: Changed the wording, what do you think?

File: Commands.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (OCHyams) [reply Commands.md]: Added DexLabel support for DexEXpectPRogramState in #40

File: README.md [hunk 6]
```diff
@@ -13,6 +13,12 @@ The following command evaluates your environment, listing the available and comp
     dexter.py list-debuggers
 
 ## Dependencies
+[TODO] Add a requirements.txt or an install.py and document it here.
+
+### Python 3.6
+
+DExTer requires at least python 3.6. This means that LLDB must also be built
+with at least 3.6 if you wish to use it.
```
Comment (jmorse) [inline README.md]: Ooof, couldnt "it" at the end refer to either Dexter or LLDB in this sentence? Better IMO to say nothing about LLDB.

File: dex/debugger/Debuggers.py [hunk 7]
```diff
@@ -216,6 +220,7 @@ def potential_debuggers(cls):
             return cls._potential_debuggers
         except AttributeError:
             cls._potential_debuggers = _get_potential_debuggers()
+            import pdb; pdb.set_trace()
```
Comment (jmorse) [inline dex/debugger/Debuggers.py]: Debug leak?

File: README.md (hunk 6 above)
Comment (OCHyams) [reply README.md]: Ambiguity is bad; removed as suggested.

**Comment by jmorse** [comment]
Tres bon; would merge but conflict in ParseCommand.py?

**Comment by OCHyams** [comment]
Fixed the conflict.

**Pull request merged**